File: scalapack/ex1.f90
!LIBPATH=/opt/intel/Compiler/11.0/081/mkl/lib/em64t
!mpif90  -O1  -o ex1  ex1.f90  \
!  $(LIBPATH)/libmkl_scalapack_lp64.a \
!  $(LIBPATH)/libmkl_intel_lp64.a \
!  $(LIBPATH)/libmkl_blacs_openmpi_lp64.a \
!  $(LIBPATH)/libmkl_sequential.a \
!  $(LIBPATH)/libmkl_core.a  \
!  -lpthread


program ples
!       Parallel Linear Equation solver
!       Array A  = ( MP*MLA x NP*NLA )
!       Proc. grid is MP     x NP
!       Local mat. (al) MLA  x    NLA
!       array A(ig,jg) = 0.1*ig + 0.001*jg  (ig not= jg)
!                      = ig                 (ig    = jg)
!       Solve Ax=b, b=1
  integer, parameter    :: MP=2,MLA=4,MB=2,NP=2,NLA=4,NB=2
  integer, parameter    :: M=MLA*MP,       N=NLA*NP
  integer, dimension(10):: idescal, idescb
  real,    allocatable  :: al(:,:), b(:)
  integer, allocatable  :: ipiv(:)
  DATA               NPROW / 2 / , NPCOL / 2 /

  CALL SL_INIT( icon, NPROW, NPCOL )

!  call blacs_gridinit( icon,'c', mp, np ) !MP & NP = 2**x
  call blacs_gridinfo( icon, mp_ret, np_ret, myrow, mycol)
!  write(*,'( "(",2i2,")"," on ",i2," x",i2," grid  ", &
!        "Al(",i3,",",i3,")" )') myrow,mycol,MP,NP,MLA,NLA
  if( (MLA/MB)*MB .ne. MLA) stop 'MLA/MB not even'
  if( (NLA/NB)*NB .ne. MLA) stop 'NLA/NB not even'

  call descinit(idescal, m,n,mb,nb, 0, 0, icon, MLA, info)
  call descinit(idescb,  m,1,nb, 1, 0, 0, icon, MLA, info)


  allocate( al(MLA,NLA), ipiv(MLA+MB), b(MLA) )
  call setarray(al,myrow,mycol,4,4)
  write(*,'(i3,i3," vals",16f10.5)')myrow,mycol,al
  b=1.0
  call PSGESV(N, 1, al, 1,1,idescal, ipiv, &
                     b, 1,1,idescb,  info)
   if(mycol .eq. 0) write(*,'( "x=(",2i2,")",4(f8.4) )' ) myrow,mycol,b
   deallocate( al, ipiv, b )
   CALL BLACS_EXIT( 0 )
end program

subroutine setarray(a,myrow,mycol,lda_x, lda_y ) 
      ! distribute a matrix aa to the local arrays.
      implicit none
      real :: aa(8,8)
      real :: bb(8,1)
      integer :: lda_x, lda_y 
      real :: a(lda_x,lda_y)
      real ::ll,mm,cr,cc
      integer :: ii,jj,i,j,myrow,mycol,pr,pc,h,g
      integer , parameter :: nprow = 2, npcol = 2
      integer, parameter ::n=8,m=8,nb=2,mb=2,rsrc=0,csrc=0
      integer, parameter :: n_b = 1 
    do i=1,8
      do j=1,8
         if(i .eq. j)then
           aa(i,j)=i
         else
           aa(i,j)=0.1*i+0.001*j
         endif
      enddo
    enddo
      do i=1,m
          do j = 1,n
      ! finding out which pe gets this i,j element
              cr = real( (i-1)/mb )
              h = rsrc+aint(cr)
              pr = mod( h,nprow)
              cc = real( (j-1)/mb )
              g = csrc+aint(cc)
              pc = mod(g,nprow)
      ! check if on this pe and then set a
              if (myrow ==pr .and. mycol==pc)then
      ! ii,jj coordinates of local array element
      ! ii = x + l*mb
      ! jj = y + m*nb
                  ll = real( ( (i-1)/(nprow*mb) ) )
                  mm = real( ( (j-1)/(npcol*nb) ) )
                  ii = mod(i-1,mb) + 1 + aint(ll)*mb
                  jj = mod(j-1,nb) + 1 + aint(mm)*nb
                  a(ii,jj) = aa(i,j)
!                  write(*,12345) i,j,ii,jj,myrow,mycol
              end if
          end do
      end do
12345 format("a(",i2,",",i2,") maps to (",i2,",",i2,") on processor ",2i3)
end subroutine

File: scalapack/ex1.c
/******
program ples
!       Parallel Linear Equation solver
!       Array A  = ( MP*MLA x NP*NLA )
!       Proc. grid is MP     x NP
!       Local mat. (al) MLA  x    NLA
!       array A(ig,jg) = 0.1*ig + 0.001*jg  (ig not= jg)
!                      = ig                 (ig    = jg)
!       Solve Ax=b, b=1
******/
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#ifdef DOMKL
#include <mkl_cblas.h>
#include <mkl_blas.h>
#include <mkl_scalapack.h>
#endif

#ifdef NOUNDER
#define PSGESV     psgesv
#define DESCINIT   descinit
#else
#define PSGESV     psgesv_
#define DESCINIT   descinit_
#endif

extern void   Cblacs_pinfo( int* mypnum, int* nprocs);
extern void   Cblacs_get( int context, int request, int* value);
extern int    Cblacs_gridinit( int* context, char * order, int np_row, int np_col);
extern void   Cblacs_gridinfo( int context, int*  np_row, int* np_col, int*  my_row, int*  my_col);
extern void   Cblacs_gridexit( int context);
extern void   Cblacs_exit( int error_code);
extern void   Cblacs_gridmap( int* context, int* map, int ld_usermap, int np_row, int np_col);

void setarray(float *a, int myrow, int mycol, int lda_x, int lda_y);
void DESCINIT(int *idescal, int *m,int *n,int *mb,int *nb, int *dummy1 , int *dummy2 , int *icon, int *mla, int *info);

	int icon;

int main() {
	int mp=2;
	int mla=4;
	int mb=2;
	int np=2;
	int nla=4;
	int nb=2;
	int mype,npe;
	int idescal[11], idescb[11];
	float *al, *b;
	int *ipiv;
	int nprow=2 , npcol=2;
	int ib;
	int info;
	int  mp_ret, np_ret, myrow, mycol;
	int zero=0;
	int one=1;
	int  m,n;
	m=mla*mp;
	n=nla*np;
	
	Cblacs_pinfo( &mype, &npe );
	Cblacs_get( -1, 0, &icon );
	Cblacs_gridinit( &icon,"c", mp, np ); /* MP & NP = 2**x */
	Cblacs_gridinfo( icon, &mp_ret, &np_ret, &myrow, &mycol);
	info=0;
	DESCINIT(idescal, &m, &n  , &mb, &nb , &zero, &zero, &icon, &mla, &info);
	info=0;
	DESCINIT(idescb,  &m, &one, &nb, &one, &zero, &zero, &icon, &mla, &info);
	
	al=(float*)malloc(sizeof(float)*mla*nla);
	b=(float*)malloc(sizeof(float)*mla);
	ipiv=(int*)malloc(sizeof(int)*(mla+mb));
	
	setarray(al,myrow,mycol,4,4);
	printf("%3d%3d vals%10.5f%10.5f%10.5f%10.5f%10.5f%10.5f%10.5f%10.5f%10.5f%10.5f%10.5f%10.5f%10.5f%10.5f%10.5f%10.5f\n",
	myrow,mycol,al[ 0],al[ 1],al[ 2],al[ 3]
	           ,al[ 4],al[ 5],al[ 6],al[ 7]
	           ,al[ 8],al[ 9],al[10],al[11]
	           ,al[12],al[13],al[14],al[15]);
	
	
	for(ib=0;ib<mla;ib++) {
		b[ib]=1.0;
	}

/*    psgesv_(n, 1, al, 1,1,idescal, ipiv, b, 1,1,idescb,  info) */
      PSGESV(&n, &one, al, &one,&one,idescal, ipiv, b, &one,&one,idescb,  &info);
      if(mycol == 0 ){
      	printf("x=(%2d%2d)%8.4f%8.4f%8.4f%8.4f\n",myrow,mycol,b[0],b[1],b[2],b[3]);
      }
	free(al);
	free(ipiv);
	free(b);
	Cblacs_exit( 0 );
return 0;	
  }
  
float **matrix(int nrl,int nrh,int ncl,int nch);
int mod(int i, int j);

void setarray(float *a,int myrow, int mycol, int lda_x, int lda_y){
      float **aa;
      float ll,mm,cr,cc;
      int ii,jj,i,j,pr,pc,h,g;
      int nprow = 2, npcol = 2;
      int n=8,m=8,nb=2,mb=2,rsrc=0,csrc=0;
      int n_b = 1;
      int index;
      aa=matrix(1,8,1,8);
      for (i=1;i<=8;i++) {
      	for (j=1;j<=8;j++) {
         if(i == j){
           aa[i][i]=i;
         }
         else {
            aa[i][j]=0.1*i+0.001*j; 
         }
         }
      }
      for (i=1;i<=m;i++) {
      	for (j=1;j<=n;j++) {
      // finding out which pe gets this i,j element
              cr = (float)( (i-1)/mb );
              h = rsrc+(int)(cr);
              pr = mod( h,nprow);
              cc = (float)( (j-1)/mb );
              g = csrc+(int)(cc);
              pc = mod(g,nprow);
      // check if on this pe and then set a
              if (myrow == pr && mycol==pc){
      // ii,jj coordinates of local array element
      // ii = x + l*mb
      // jj = y + m*nb
                  ll = (float)( ( (i-1)/(nprow*mb) ) );
                  mm = (float)( ( (j-1)/(npcol*nb) ) );
                  ii = mod(i-1,mb) + 1 + (int)(ll)*mb;
                  jj = mod(j-1,nb) + 1 + (int)(mm)*nb;
                  index=(jj-1)*lda_x+ii;
                  index=index-1;
/*                  a(ii,jj) = aa(i,j) */
                  a[index] = aa[i][j];
              }
          }
		}

  	}

int mod(int i, int j) {
	return (i % j);
}
float **matrix(int nrl,int nrh,int ncl,int nch)
{
    int i;
	float **m;
	m=(float **) malloc((unsigned) (nrh-nrl+1)*sizeof(float*));
	if (!m){
	     printf("allocation failure 1 in matrix()\n");
	     exit(1);
	}
	m -= nrl;
	for(i=nrl;i<=nrh;i++) {
	    if(i == nrl){ 
		    m[i]=(float *) malloc((unsigned) (nrh-nrl+1)*(nch-ncl+1)*sizeof(float));
		    if (!m[i]){
		         printf("allocation failure 2 in matrix()\n");
		         exit(1);
		    }		
		    m[i] -= ncl;
	    }
	    else {
	        m[i]=m[i-1]+(nch-ncl+1);
	    }
	}
	return m;
}

